criu/files-reg.c

```diff
@@ -1581,6 +1587,110 @@ static int get_build_id(const int fd, const struct stat *fd_status,
 	return ret;
 }
 
+static inline void calculate_checksum_iterator_init(u64 *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FULL)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(u64 *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FULL)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter += opts.file_validation_chksm_parameter;
+}
+
+static inline bool calculate_checksum_iterator_stop_condition(u64 iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FULL) {
+		return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		if (iter >= opts.file_validation_chksm_parameter)
+			return true;
+		else
+			return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Does the actual work of calculating the CRC32C checksum of "some"
+ * parts of the file. These "some" parts depend on the configuration chosen.
+ * By default, the first 1024 bytes of the file or the entire file, which ever
+ * is smaller, is considered.
+ * Returns true if the checksum could be obtained, else it returns false.
+ */
+static bool calculate_checksum(const int fd, const struct stat *fd_status,
+				u32 *checksum)
+{
+	int tmp;
+	u32 byte, mask;
+	u64 up_bound, low_bound = 0, i = 0;
+	unsigned char *file_header;
+
+	/* At most, the first 10MB of the file is mapped. */
+	up_bound = min_t(size_t, fd_status->st_size, CHKSM_CHUNK_SIZE);
+	file_header = (unsigned char *) mmap(0, up_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (file_header == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return false;
+	}
+
+	*checksum = 0xFFFFFFFF;
+	calculate_checksum_iterator_init(&i);
+
+	while (!calculate_checksum_iterator_stop_condition(i) &&
+			i >= low_bound && i < up_bound) {
+
+		byte = file_header[i-low_bound];
+		*checksum = *checksum ^ byte;
+		for (tmp = 0; tmp < 8; tmp++) {
+			mask = -(*checksum & 1);
```

Comment:
How fast is this implementation of crc32? Have you looked at how it is implemented in the Linux kernel or other projects?

Reply:
I don't think you can make it any faster (Unrolling the for loop would help a bit I guess) without using a lookup table, which is what the Linux kernel does.

Reply:
why do we not use the lookup table?

Reply:
Since the checksum method is more like a fallback for the build-id method, I thought it seemed unnecessary. The lookup table is a necessity for the Linux kernel since it has to run for every network packet sent and received.

Regardless of this, I've mentioned it under "Possible Improvements" in the document I sent you for the wiki page.

Reply:
Do we do validations for all files (not only mapped files)? I think the answer is yes and it isn't just a fallback in this case. Right now, this checksum code is very slow. We decided that you don't need to fix it right now. If you will have a wish, you can rework it after the GSOC program.